File: iOS/Athlynk/Athlynk/Features/CheckHistoryDetailView.swift
//
//  CheckHistoryDetailView.swift
//  Read-only detail of one past submitted check: every answered question,
//  section by section — including the coach-only «Calcolo Fabbisogni» tool,
//  which the athlete never fills but can review. Mirrors WebApp check/detail.html.
//  Backed by GET /api/v1/checks/<id>.
//

import SwiftUI

struct CheckHistoryDetailView: View {
    let responseId: Int

    @State private var data: CheckDetailDTO?
    @State private var loading = true
    @State private var error: String?

    var body: some View {
        ZStack {
            VoltBackground(palette: [Palette.violet, Palette.cyan, Palette.magenta, Palette.violet])
            ScrollView(showsIndicators: false) {
                VStack(alignment: .leading, spacing: 18) {
                    if loading {
                        ProgressSkeleton()
                    } else if let error {
                        EmptyPanel(icon: "wifi.exclamationmark", text: error, color: Palette.danger)
                    } else if let d = data {
                        header(d)
                        ForEach(d.sections) { section(for: $0) }
                        if !d.photos.isEmpty { photos(d) }
                        if hasNotes(d) { notes(d) }
                    }
                }
                .padding(.horizontal, 22).padding(.top, 12).padding(.bottom, 40)
            }
        }
        .navigationTitle(data?.title ?? "Check")
        .navigationBarTitleDisplayMode(.inline)
        .toolbarBackground(.hidden, for: .navigationBar)
        .task { await load() }
    }

    private func header(_ d: CheckDetailDTO) -> some View {
        VStack(alignment: .leading, spacing: 4) {
            Text("CHECK COMPILATO").font(Typo.mono(10, .bold)).tracking(3).foregroundStyle(Palette.violet)
            Text(dateLong(parseDate(d.submittedAt))).font(Typo.body(14)).foregroundStyle(Palette.textMid)
        }
    }

    // MARK: Sections

    @ViewBuilder private func section(for s: CheckSection) -> some View {
        if !s.questions.isEmpty {
            VStack(alignment: .leading, spacing: 10) {
                Text(s.label.uppercased()).voltEyebrow()
                ForEach(s.questions) { q in questionRow(q) }
            }
        }
    }

    @ViewBuilder private func questionRow(_ q: CheckSectionQuestion) -> some View {
        switch q.type {
        case "strumento_fabbisogni":
            if let fb = q.fb { fabbisogniCard(fb) }
        case "allegato":
            if let files = q.files, !files.isEmpty { attachmentRow(q.label, files) }
        case "metrica_pair":
            if let sides = q.sides { pairRow(q.label, sides) }
        default:
            if let display = q.value?.display {
                simpleRow(q.label, display, unit: q.unit, previous: q.previous?.display, delta: q.delta)
            }
        }
    }

    private func simpleRow(_ label: String, _ value: String, unit: String?, previous: String?, delta: Double?) -> some View {
        HStack(alignment: .top) {
            Text(label).font(Typo.body(14)).foregroundStyle(Palette.textMid)
            Spacer()
            VStack(alignment: .trailing, spacing: 2) {
                HStack(spacing: 4) {
                    Text(value).font(Typo.mono(14, .bold)).foregroundStyle(Palette.textHi)
                    if let unit { Text(unit).font(Typo.mono(10)).foregroundStyle(Palette.textLow) }
                }
                if let delta, delta != 0 {
                    Text("\(delta > 0 ? "+" : "")\(String(format: "%.1f", delta))")
                        .font(Typo.mono(10, .semibold)).foregroundStyle(delta > 0 ? Palette.lime : Palette.magenta)
                }
            }
        }
        .padding(.horizontal, 14).padding(.vertical, 12).voltPanel(radius: 12)
    }

    private func pairRow(_ label: String, _ sides: [CheckMeasureSide]) -> some View {
        HStack {
            Text(label).font(Typo.body(14)).foregroundStyle(Palette.textMid)
            Spacer()
            HStack(spacing: 14) {
                ForEach(Array(sides.enumerated()), id: \.offset) { _, side in
                    VStack(alignment: .trailing, spacing: 1) {
                        if let tag = side.tag { Text(tag).font(Typo.mono(9, .bold)).foregroundStyle(Palette.textLow) }
                        HStack(spacing: 3) {
                            Text(side.value?.display ?? "—").font(Typo.mono(13, .bold)).foregroundStyle(Palette.textHi)
                            if let unit = side.unit { Text(unit).font(Typo.mono(9)).foregroundStyle(Palette.textLow) }
                        }
                    }
                }
            }
        }
        .padding(.horizontal, 14).padding(.vertical, 12).voltPanel(radius: 12)
    }

    private func attachmentRow(_ label: String, _ files: [CheckAttachmentFile]) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(label).font(Typo.body(13)).foregroundStyle(Palette.textMid)
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 10) {
                    ForEach(files) { f in attachmentThumb(f) }
                }
            }
        }
        .padding(14).voltPanel(radius: 12)
    }

    private func attachmentThumb(_ f: CheckAttachmentFile) -> some View {
        let isImage = (f.mimeType ?? "").hasPrefix("image") || f.url.lowercased().hasSuffix(".jpg")
            || f.url.lowercased().hasSuffix(".jpeg") || f.url.lowercased().hasSuffix(".png")
        return Link(destination: URL(string: f.url) ?? URL(string: "about:blank")!) {
            if isImage {
                AsyncImage(url: URL(string: f.url)) { phase in
                    switch phase {
                    case .success(let img): img.resizable().scaledToFill()
                    default: Rectangle().fill(Palette.void2)
                    }
                }
                .frame(width: 88, height: 88)
                .clipShape(RoundedRectangle(cornerRadius: 10))
            } else {
                VStack(spacing: 4) {
                    Image(systemName: "doc.fill").font(.system(size: 22)).foregroundStyle(Palette.violet)
                    Text(f.fileName).font(Typo.mono(8)).foregroundStyle(Palette.textLow).lineLimit(1)
                }
                .frame(width: 88, height: 88)
                .background(RoundedRectangle(cornerRadius: 10).fill(Palette.void2))
            }
        }
    }

    // MARK: Calcolo Fabbisogni (read-only — mirrors WebApp check/detail.html)

    private func fabbisogniCard(_ fb: CheckFabbisogni) -> some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack(spacing: 6) {
                Image(systemName: "function").foregroundStyle(Palette.bronze)
                Text("Calcolo Fabbisogni").font(Typo.body(13, .semibold)).foregroundStyle(Palette.bronze)
            }

            if let det = fb.detFinale {
                HStack(alignment: .firstTextBaseline) {
                    VStack(alignment: .leading, spacing: 2) {
                        Text("DET FINALE").font(Typo.mono(9, .bold)).foregroundStyle(Palette.textLow)
                        (Text("\(det)").font(Typo.poster(24)) + Text(" kcal/die").font(Typo.body(12)))
                            .foregroundStyle(Palette.textHi)
                    }
                    Spacer()
                    if fb.detAdjust != 0, let base = fb.detBase {
                        StatusBadge(text: "base \(base) · \(fb.detAdjust > 0 ? "+" : "")\(fb.detAdjust) kcal", color: Palette.bronze)
                    }
                }
                .padding(14).voltPanel(radius: 12)
            }

            let stats: [(String, String)] = [
                fb.peso?.display.map { ("Peso", "\($0) kg") },
                fb.altezza?.display.map { ("Altezza", "\($0) cm") },
                fb.eta?.display.map { ("Età", $0) },
                fb.sesso?.display.map { ("Sesso", $0) },
                fb.mb.map { ("MB", "\($0) kcal") },
                fb.pal?.display.map { ("PAL", "\($0)×") },
                fb.formula?.display.map { ("Formula", $0) },
            ].compactMap { $0 }
            if !stats.isEmpty {
                LazyVGrid(columns: [GridItem(.flexible()), GridItem(.flexible())], spacing: 8) {
                    ForEach(stats, id: \.0) { r in
                        HStack {
                            Text(r.0).font(Typo.body(12)).foregroundStyle(Palette.textMid)
                            Spacer()
                            Text(r.1).font(Typo.mono(12, .bold)).foregroundStyle(Palette.textHi)
                        }
                        .padding(.horizontal, 12).padding(.vertical, 10).voltPanel(radius: 10)
                    }
                }
            }

            if !fb.macros.isEmpty {
                VStack(spacing: 0) {
                    ForEach(Array(fb.macros.enumerated()), id: \.offset) { idx, m in
                        HStack {
                            Text(m.label).font(Typo.body(13, .semibold)).foregroundStyle(Palette.textHi)
                            Spacer()
                            let parts = [m.gkg?.display.map { "\($0) g/kg" }, m.g.map { "\($0) g" }, m.kcal.map { "\($0) kcal" }]
                                .compactMap { $0 }.joined(separator: " · ")
                            Text(parts).font(Typo.mono(12)).foregroundStyle(Palette.textMid)
                        }
                        .padding(.horizontal, 14).padding(.vertical, 9)
                        if idx < fb.macros.count - 1 { Divider().opacity(0.3) }
                    }
                }
                .voltPanel(radius: 12)
            }

            let extras: [(String, String)] = [
                fb.fibra?.display.map { ("Fibra", "\($0) g/die") },
                fb.idrico?.display.map { ("Idratazione", "\($0) ml/die") },
            ].compactMap { $0 }
            if !extras.isEmpty {
                HStack(spacing: 16) {
                    ForEach(extras, id: \.0) { r in
                        VStack(alignment: .leading, spacing: 2) {
                            Text(r.0.uppercased()).font(Typo.mono(9, .bold)).foregroundStyle(Palette.textLow)
                            Text(r.1).font(Typo.body(13, .semibold)).foregroundStyle(Palette.textHi)
                        }
                    }
                }
            }
            fbNote("Attività", fb.palDesc?.display)
            fbNote("Note DET", fb.detNote?.display)
            fbNote("Micronutrienti", fb.micro?.display)
            fbNote("Note operative", fb.noteOp?.display)
        }
        .padding(14).voltPanel(Palette.bronze.opacity(0.3), radius: 14)
    }

    @ViewBuilder private func fbNote(_ label: String, _ value: String?) -> some View {
        if let value, !value.isEmpty {
            VStack(alignment: .leading, spacing: 2) {
                Text(label.uppercased()).font(Typo.mono(9, .bold)).foregroundStyle(Palette.textLow)
                Text(value).font(Typo.body(13)).foregroundStyle(Palette.textHi)
            }
        }
    }

    // MARK: Photos / notes

    private func photos(_ d: CheckDetailDTO) -> some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("FOTO PROGRESSI").voltEyebrow()
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 12) {
                    ForEach(d.photos) { p in
                        AsyncImage(url: URL(string: p.url)) { phase in
                            switch phase {
                            case .success(let img): img.resizable().scaledToFill()
                            default: Rectangle().fill(Palette.void2)
                            }
                        }
                        .frame(width: 130, height: 175)
                        .clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
                    }
                }
            }
        }
    }

    private func notes(_ d: CheckDetailDTO) -> some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("NOTE & FEEDBACK").voltEyebrow()
            if let fb = d.coachFeedback.isEmpty ? nil : d.coachFeedback {
                HStack(alignment: .top, spacing: 8) {
                    Image(systemName: "quote.opening").font(.system(size: 12, weight: .black)).foregroundStyle(Palette.cyan)
                    Text(fb).font(Typo.body(13)).foregroundStyle(Palette.textMid)
                }
                .padding(12).background(RoundedRectangle(cornerRadius: 10).fill(Palette.void2))
            }
            fbNote("Messaggio", d.notes)
            fbNote("Infortuni", d.injuries)
            fbNote("Limitazioni", d.limitations)
        }
    }

    private func hasNotes(_ d: CheckDetailDTO) -> Bool {
        !d.coachFeedback.isEmpty || [d.notes, d.injuries, d.limitations].contains { ($0 ?? "").isEmpty == false }
    }

    // MARK: Load / helpers

    private func load() async {
        loading = true; defer { loading = false }
        do { data = try await APIClient.shared.checkDetail(responseId: responseId) }
        catch { self.error = "Impossibile caricare il check." }
    }

    private func parseDate(_ iso: String?) -> Date? { iso.flatMap(ISO8601.parse) }
    private func dateLong(_ d: Date?) -> String {
        guard let d else { return "—" }
        let f = DateFormatter(); f.locale = Locale(identifier: "it_IT"); f.dateFormat = "d MMMM yyyy"
        return f.string(from: d)
    }
}

File: android/lib/features/athlete/progress/check_history_detail_view.dart
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/api/athlete_api.dart';
import '../../../core/l10n/formatters.dart';
import '../../../core/models/models.dart';
import '../../../core/providers.dart';
import '../../../design/components/panel.dart';
import '../../../design/components/scaffold.dart';
import '../../../design/components/skeleton.dart';
import '../../../design/theme.dart';

/// Read-only past submitted check — port of iOS `CheckHistoryDetailView`:
/// per-section questions (value+delta, paired DX/SX, attachments, and the
/// read-only Calcolo Fabbisogni summary), photos, notes, coach feedback.
class CheckHistoryDetailView extends ConsumerStatefulWidget {
  const CheckHistoryDetailView({super.key, required this.responseId});

  final int responseId;

  @override
  ConsumerState<CheckHistoryDetailView> createState() =>
      _CheckHistoryDetailViewState();
}

class _CheckHistoryDetailViewState
    extends ConsumerState<CheckHistoryDetailView> {
  CheckDetailDto? _detail;
  bool _error = false;

  @override
  void initState() {
    super.initState();
    _load();
  }

  Future<void> _load() async {
    try {
      final d =
          await ref.read(apiClientProvider).checkDetail(widget.responseId);
      if (mounted) setState(() => _detail = d);
    } catch (_) {
      if (mounted) setState(() => _error = true);
    }
  }

  @override
  Widget build(BuildContext context) {
    final d = _detail;
    return Scaffold(
      backgroundColor: Colors.transparent,
      appBar: AppBar(),
      body: d == null
          ? Padding(
              padding: const EdgeInsets.all(Space.screenH),
              child: _error
                  ? EmptyPanel.network(onCta: () {
                      setState(() => _error = false);
                      _load();
                    })
                  : const FormSkeleton(),
            )
          : ScreenScroll(
              topPadding: 0,
              spacing: Space.element,
              children: [
                ScreenHeader(
                  eyebrow: 'Check inviato',
                  title: d.title,
                  titleSize: 30,
                  subtitle: Formatters.parseDate(d.submittedAt) == null
                      ? null
                      : Formatters.longDate(
                          Formatters.parseDate(d.submittedAt)!),
                ),
                for (final section in d.sections) _sectionCard(section),
                if (d.photos.isNotEmpty) _photos(d.photos),
                if ((d.notes ?? '').isNotEmpty)
                  _noteCard('Note', d.notes!, Palette.cyan),
                if ((d.injuries ?? '').isNotEmpty)
                  _noteCard('Infortuni', d.injuries!, Palette.crimson),
                if ((d.limitations ?? '').isNotEmpty)
                  _noteCard('Limitazioni', d.limitations!, Palette.amber),
                if (d.coachFeedback.isNotEmpty)
                  VoltPanel(
                    tint: Palette.amber.withValues(alpha: 0.45),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        const Eyebrow('Feedback del coach'),
                        const SizedBox(height: 8),
                        Text(d.coachFeedback,
                            style: Typo.body(14.5, FontWeight.w500)),
                      ],
                    ),
                  ),
              ],
            ),
    );
  }

  Widget _sectionCard(CheckSection section) {
    return VoltPanel(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Eyebrow(section.label),
          const SizedBox(height: 10),
          for (final q in section.questions) _question(q),
        ],
      ),
    );
  }

  Widget _question(CheckSectionQuestion q) {
    if (q.fb != null) return _fabbisogniCard(q.fb!);
    if (q.sides != null && q.sides!.isNotEmpty) {
      return Padding(
        padding: const EdgeInsets.only(bottom: 10),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(q.label,
                style: Typo.body(13.5, FontWeight.w600, Palette.textMid)),
            const SizedBox(height: 6),
            Row(
              children: [
                for (final side in q.sides!)
                  Expanded(
                    child: Container(
                      margin: const EdgeInsets.only(right: 8),
                      padding: const EdgeInsets.all(10),
                      decoration: BoxDecoration(
                        color: Palette.void0,
                        borderRadius: BorderRadius.circular(10),
                        border: Border.all(color: Palette.line),
                      ),
                      child: Column(
                        children: [
                          Text(side.tag ?? side.label,
                              style: Typo.mono(
                                  9, FontWeight.w700, Palette.textLow)),
                          const SizedBox(height: 4),
                          Text(
                            '${side.value?.display ?? "—"}${side.unit == null ? '' : ' ${side.unit}'}',
                            style: Typo.mono(15, FontWeight.w700),
                          ),
                          if (side.delta != null)
                            Text(
                              '${side.delta! >= 0 ? '+' : ''}${Formatters.decimal(side.delta!)}',
                              style: Typo.mono(
                                  10,
                                  FontWeight.w600,
                                  side.delta! == 0
                                      ? Palette.textLow
                                      : (side.delta! > 0
                                          ? Palette.lime
                                          : Palette.crimson)),
                            ),
                        ],
                      ),
                    ),
                  ),
              ],
            ),
          ],
        ),
      );
    }
    if (q.files != null && q.files!.isNotEmpty) {
      return Padding(
        padding: const EdgeInsets.only(bottom: 10),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(q.label,
                style: Typo.body(13.5, FontWeight.w600, Palette.textMid)),
            const SizedBox(height: 6),
            SizedBox(
              height: 74,
              child: ListView(
                scrollDirection: Axis.horizontal,
                children: [
                  for (final f in q.files!)
                    Padding(
                      padding: const EdgeInsets.only(right: 8),
                      child: ClipRRect(
                        borderRadius: BorderRadius.circular(10),
                        child: CachedNetworkImage(
                          imageUrl: f.url,
                          width: 58,
                          height: 74,
                          fit: BoxFit.cover,
                          errorWidget: (_, _, _) => Container(
                              width: 58,
                              color: Palette.void2,
                              child: const Icon(Icons.insert_drive_file_outlined,
                                  size: 18, color: Palette.textLow)),
                        ),
                      ),
                    ),
                ],
              ),
            ),
          ],
        ),
      );
    }
    return Padding(
      padding: const EdgeInsets.only(bottom: 8),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Expanded(
            child: Text(q.label,
                style: Typo.body(13.5, FontWeight.w500, Palette.textMid)),
          ),
          const SizedBox(width: 10),
          Column(
            crossAxisAlignment: CrossAxisAlignment.end,
            children: [
              Text(
                '${q.value?.display ?? "—"}${q.unit == null ? '' : ' ${q.unit}'}',
                style: Typo.body(14, FontWeight.w700),
              ),
              if (q.delta != null && q.delta != 0)
                Text(
                  '${q.delta! > 0 ? '+' : ''}${Formatters.decimal(q.delta!)}',
                  style: Typo.mono(10, FontWeight.w600,
                      q.delta! > 0 ? Palette.lime : Palette.crimson),
                ),
            ],
          ),
        ],
      ),
    );
  }

  Widget _fabbisogniCard(CheckFabbisogni fb) {
    Widget row(String label, String? value) => value == null || value.isEmpty
        ? const SizedBox.shrink()
        : Padding(
            padding: const EdgeInsets.symmetric(vertical: 2),
            child: Row(
              children: [
                Expanded(
                    child: Text(label,
                        style: Typo.body(
                            13, FontWeight.w500, Palette.textMid))),
                Text(value, style: Typo.body(13, FontWeight.w700)),
              ],
            ),
          );

    return Container(
      margin: const EdgeInsets.only(bottom: 10),
      padding: const EdgeInsets.all(12),
      decoration: BoxDecoration(
        color: Palette.void0,
        borderRadius: BorderRadius.circular(12),
        border: Border.all(color: Palette.goldText.withValues(alpha: 0.4)),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          const Eyebrow('Calcolo Fabbisogni'),
          const SizedBox(height: 8),
          row('Metabolismo basale', fb.mb?.toString()),
          row('Formula', fb.formula?.display),
          row('PAL', fb.pal?.display),
          row('DET finale', fb.detFinale == null ? null : '${fb.detFinale} kcal'),
          if (fb.macros.isNotEmpty) ...[
            const SizedBox(height: 8),
            for (final m in fb.macros)
              Padding(
                padding: const EdgeInsets.symmetric(vertical: 2),
                child: Row(
                  children: [
                    Expanded(
                        child: Text(m.label,
                            style: Typo.body(13, FontWeight.w600))),
                    Text(
                      '${m.g ?? "—"} g · ${m.kcal ?? "—"} kcal',
                      style: Typo.mono(12, FontWeight.w600, Palette.textMid),
                    ),
                  ],
                ),
              ),
          ],
          row('Fibra', fb.fibra?.display),
          row('Apporto idrico', fb.idrico?.display),
        ],
      ),
    );
  }

  Widget _photos(List<CheckDetailPhoto> photos) {
    return VoltPanel(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          const Eyebrow('Foto progresso'),
          const SizedBox(height: 10),
          SizedBox(
            height: 110,
            child: ListView(
              scrollDirection: Axis.horizontal,
              children: [
                for (final p in photos)
                  Padding(
                    padding: const EdgeInsets.only(right: 8),
                    child: ClipRRect(
                      borderRadius: BorderRadius.circular(12),
                      child: CachedNetworkImage(
                        imageUrl: p.url,
                        width: 84,
                        height: 110,
                        fit: BoxFit.cover,
                        errorWidget: (_, _, _) =>
                            Container(width: 84, color: Palette.void2),
                      ),
                    ),
                  ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  Widget _noteCard(String title, String body, Color accent) {
    return VoltPanel(
      tint: accent.withValues(alpha: 0.35),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Eyebrow(title),
          const SizedBox(height: 6),
          Text(body, style: Typo.body(14, FontWeight.w400)),
        ],
      ),
    );
  }
}
